### PR TITLE
Update color assertions

### PR DESCRIPTION
**Pull request by gaaclarke**
issue: https://github.com/flutter/flutter/issues/127855
This is a forward fix to help https://github.com/flutter/engine/pull/54981 land.

It makes the following changes:
1) Removes hardcoded `Color.toString()` assumptions in asserts
1) Switches some lerp tests to use numbers that are more friendly to uint8_t and floating point numbers
1) implements custom matchers for color
1) removes word wrapping for some asserts since it makes them fragile to changes in `Color.toString()` invocations

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/test/painting/border_rtl_test.dart [hunk 1]
```diff
@@ -13,6 +13,69 @@ class SillyBorder extends BoxBorder {
...
+  @override
+  bool matches(dynamic item, Map<dynamic, dynamic> matchState) {
+    if (item is Border && _target is Border) {
+      return _sideMatches(item.top, _target.top) &&
+          _sideMatches(item.right, _target.right) &&
+          _sideMatches(item.bottom, _target.bottom) &&
+          _sideMatches(item.left, _target.left);
+    } else {
+      return false;
+    }
```
Comment (goderbauer) [inline packages/flutter/test/painting/border_rtl_test.dart:47]: Couldn't this just be simplified to:

```dart
return item is Border && _target is Border &&
    _sideMatches(item.top, _target.top) &&
    _sideMatches(item.right, _target.right) &&
    _sideMatches(item.bottom, _target.bottom) &&
    _sideMatches(item.left, _target.left);
```

File: packages/flutter/test/painting/border_rtl_test.dart [hunk 2]
```diff
@@ -13,6 +13,69 @@ class SillyBorder extends BoxBorder {
...
+
+  @override
+  Description describe(Description description) {
+    description.add('expected $_target');
+    return description;
+  }
+
+  @override
+  bool matches(dynamic item, Map<dynamic, dynamic> matchState) {
+    if (item is BorderDirectional && _target is BorderDirectional) {
```
Comment (goderbauer) [inline packages/flutter/test/painting/border_rtl_test.dart:66]: same here

File: packages/flutter/test/painting/gradient_test.dart [hunk 3]
```diff
@@ -12,6 +12,128 @@ import 'dart:math' as math;
...
+
+  @override
+  Description describe(Description description) {
+    description.add('expected $_target');
+    return description;
+  }
+
+  @override
+  bool matches(dynamic item, Map<dynamic, dynamic> matchState) {
+    if (item is LinearGradient) {
```
Comment (goderbauer) [inline packages/flutter/test/painting/gradient_test.dart:61]: same

File: packages/flutter_test/lib/src/matchers.dart [hunk 4]
```diff
@@ -222,9 +222,11 @@ const Matcher isNotInCard = _IsNotInCard();
 
 /// Asserts that the object represents the same color as [color] when used to paint.
 ///
-/// Specifically this matcher checks the object is of type [Color] and its [Color.value]
-/// equals to that of the given [color].
-Matcher isSameColorAs(Color color) => _ColorMatcher(targetColor: color);
+/// Specifically this matcher checks the object is of type [Color] and its color
+/// components fall below the specified delta specified by [threshold].
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/matchers.dart:226]: maybe drop one of the "specified"s?

```suggestion
/// components fall below the delta specified by [threshold].
```

File: packages/flutter_test/lib/src/matchers.dart [hunk 5]
```diff
@@ -2117,22 +2119,28 @@ class _CoversSameAreaAs extends Matcher {
...
-    }
-    return false;
+  Description describe(Description description) {
+    return description.add('matches color "$_target" with threshold "$_threshold".');
   }
 
   @override
-  Description describe(Description description) => description.add('matches color $targetColor');
+  bool matches(dynamic item, Map<dynamic, dynamic> matchState) {
+    if (item is ui.Color) {
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/matchers.dart:2134]: same

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 6]
```diff
@@ -1768,12 +1768,13 @@ abstract class DiagnosticsNode {
     String? prefixOtherLines,
     TextTreeConfiguration? parentConfiguration,
     DiagnosticLevel minLevel = DiagnosticLevel.debug,
+    int wrapWidth = 65,
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply packages/flutter/lib/src/foundation/diagnostics.dart:1774]: done

File: packages/flutter/test/painting/border_rtl_test.dart (hunk 1 above)
Comment (gaaclarke) [reply packages/flutter/test/painting/border_rtl_test.dart]: done

File: packages/flutter/test/painting/gradient_test.dart (hunk 3 above)
Comment (gaaclarke) [reply packages/flutter/test/painting/gradient_test.dart]: done

File: packages/flutter_test/lib/src/matchers.dart (hunk 4 above)
Comment (gaaclarke) [reply packages/flutter_test/lib/src/matchers.dart]: done

File: packages/flutter_test/lib/src/matchers.dart (hunk 5 above)
Comment (gaaclarke) [reply packages/flutter_test/lib/src/matchers.dart]: done

File: packages/flutter/test/painting/border_rtl_test.dart (hunk 2 above)
Comment (gaaclarke) [reply packages/flutter/test/painting/border_rtl_test.dart]: done

**Pull request merged**